### ISSUE TITLE
add full infra-as-code support for all experience assets

### ISSUE THREAD

**Issue by blake-mealey** [blake-mealey/mantle#11]
(DRAFT)

~Experience icon:~
* ~Create: POST https://publish.roblox.com/v1/games/{gameId}/icon~ added in #12 

~Experience thumbnails:~
* ~Create: POST https://publish.roblox.com/v1/games/{gameId}/thumbnail/image~ added in #12 
* ~Change order: POST https://develop.roblox.com/v1/universes/{universeId}/thumbnails/order~ added in #12 
* ~Delete: DELETE https://develop.roblox.com/v1/universes/{universeId}/thumbnails/{thumbnailId}~ added in #12 

~Experience developer products:~
* ~Create: POST https://develop.roblox.com/v1/universes/{universeId}/developerproducts~ added in #32 
  * ~Create does not return the ID required to update. The ID can be retrieved via the list endpoint:~ added in #32 
    * ~GET https://api.roblox.com/developerproducts?universeId={}&page={}~ added in #32 
* ~Update: POST https://develop.roblox.com/v1/universes/{universeId}/developerproducts/{productId}/update~ added in #32 
* ~No way to delete products~ added in #32 
* ~Icons are uploaded separately and their asset ID is passed to create/update:~ added in #44 
  * ~Create: POST https://www.roblox.com/places/developerproduct-icon~ added in #44 

~Experience badges:~ added in #59 
* ~Create: POST https://badges.roblox.com/v1/universes/{universeId}/badges~
* ~Update config: PATCH https://badges.roblox.com/v1/badges/{badgeId}~
* ~Update icon: POST https://publish.roblox.com/v1/badges/{badgeId}/icon~
* ~No way to delete badges, but they can be marked as disabled with the update endpoint~
* ~Costs Robux so there should be a warning or something before creating~

~Experience game passes:~ added in #45 
* ~Create: POST https://www.roblox.com/build/doverifiedupload~ added in #45 
* ~Update config: POST https://www.roblox.com/game-pass/update~ added in #45 
* ~Update icon: POST https://publish.roblox.com/v1/game-passes/{gamePassId}/icon~ added in #45 
* ~No way to delete game passes~ added in #45 

~Experiences:~
* ~Create (unverified): POST https://api.roblox.com/universes/create~ added in #39 
* ~Update config: PATCH https://develop.roblox.com/v2/universes/{universeId}/configuration~ added in #12
* ~Delete: reconfigure to set `isArchived: true`~ added in #39 

~Places:~
* ~Create: POST https://www.roblox.com/ide/places/createV2~ added in #39 
* ~Update config: PATCH https://develop.roblox.com/v2/places/{placeId}~ added in #12
* ~Update file (Open Cloud): POST https://apis.roblox.com/universes/v1/{experienceId}/places/{placeId}/versions~ added in #12
* ~Update file (old): POST https://data.roblox.com/Data/Upload.ashx~ added in #39 
* ~Delete: POST https://www.roblox.com/universes/removeplace~ added in #39

Assets (images, audio, etc):
* See comment below

**Comment by blake-mealey** [comment]
Reference for creating experiences:

```js
axios({
            url: `https://api.roblox.com/universes/create?groupId=${groupId}`,
            method: "POST",
            headers: {
                Cookie: `.ROBLOSECURITY=${cookie}`,
                "User-Agent": "RobloxStudio/WinInet RobloxApp/0.461.0.416397 (GlobalDist; RobloxDirectDownload)",
                "X-CSRF-TOKEN": xcsrfToken || "",
            },
            data: {
                templatePlaceIdToUse: 95206881
            }
        })
```

**Comment by blake-mealey** [comment]
Game assets should be supported as well, specifically: images, audio, and meshes.

Game assets are created via two steps. First, they are uploaded to Roblox, then they are assigned an alias which can then be used by the [rbxgameasset content data type](https://developer.roblox.com/en-us/articles/Content#rbxgameasset).

Aliases:
* Create: POST https://develop.roblox.com/v1/universes/{universeId}/aliases
* Update: PATCH https://develop.roblox.com/v1/universes/{universeId}/aliases/{name}
* Delete: DELETE https://develop.roblox.com/v1/universes/{universeId}/aliases/{name}

Images:
* Create: POST https://data.roblox.com/data/upload/json?assetTypeId=13&name={}&description={}
* Update: Re-create and update the alias to point to the new asset
* Delete: Delete the alias

Audio:
* Estimate cost: POST https://publish.roblox.com/v1/audio/verify
  * Requires: name, duration (measured in seconds), and file size (measured in bytes)
* Create: POST https://publish.roblox.com/v1/audio
  * Requires: name (Audio/<name>), file (Base64 encoded)

Configuration API:

```yml
templates;
  # or should we split them up by type?
  assets:
    # just a file path, uses file name without extension for alias
    - file/path
    # file paths with glob support
    - file/path/*
    # file path with custom alias, does not support globs
    - name: name
      path: file/path
```

**Comment by blake-mealey** [comment]
Does it make sense to support animations? I am not sure how that work (imported from .rbxmx files maybe?)

**Comment by blake-mealey** [comment]
Closing in favour of #122 as the rest of the resources have been added
